utils/pattern/src/interpolator/mod.rs

```diff
@@ -0,0 +1,400 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+mod error;
+use crate::{replacement::ReplacementProvider, token::PatternToken};
+pub use error::InterpolatorError;
+use std::{
+    borrow::Cow,
+    fmt::{Debug, Display, Formatter},
+    str::FromStr,
+};
+use writeable::Writeable;
+
+/// The type returned by the [`Interpolator`] iterator.
+/// This enum stores references to string literals parsed as
+/// part of the pattern and elements returned by the [`ReplacementProvider`].
+///
+/// # Lifetimes
+///
+/// - `i`: The life time of a kind that is being interpolated.
+/// - `s`: The life time of a string slice literal.
+#[derive(Debug, PartialEq)]
+pub enum InterpolatedKind<'i, 's, E> {
+    Literal(&'i Cow<'s, str>),
+    Element(&'i E),
+}
+
+impl<E> PartialEq<E> for InterpolatedKind<'_, '_, E>
+where
+    E: PartialEq,
+{
+    fn eq(&self, other: &E) -> bool {
+        if let Self::Element(element) = self {
```

Comment:
nit: add a comment explaining why the Eq impl does this

Reply:
The way the application of this feedback went felt movie-like:

I started by eagerly typing "We provide a custom Eq implementation because... because..." - 10 min later I removed the block.

We may add it later (I now know why I added it!), but it's not needed now so removing. Thank you:)